presto-main/src/test/java/com/facebook/presto/sql/planner/iterative/TestMemo.java

```diff
@@ -0,0 +1,250 @@
+
+        assertMatchesStructure(
+                memo.extract(),
+                project(x.getId(),
+                        filter(y.getId(),
+                                values(z.getId()))));
+    }
+
+    /*
+      From: X -> Y -> Z
+      To:   X --> Y1' --> Z
+              \-> Y2' -/
+     */
+    @Test
+    public void testMultipleReferences()
+            throws Exception
+    {
+        PlanNode z = values();
+        PlanNode y = filter(z);
+        PlanNode x = project(y);
+
+        Memo memo = new Memo(idAllocator, x);
+        assertEquals(memo.getGroupCount(), 3);
+
+        int yGroup = getChildGroup(memo, memo.getRootGroup());
+
+        PlanNode rewrittenZ = memo.getNode(yGroup).getSources().get(0);
+        PlanNode y1 = filter(rewrittenZ);
+        PlanNode y2 = project(rewrittenZ);
+
+        PlanNode newX = join(y1, y2);
+        memo.replace(memo.getRootGroup(), newX, "rule");
+        assertEquals(memo.getGroupCount(), 4);
+
+        assertMatchesStructure(
+                memo.extract(),
+                join(newX.getId(),
+                        filter(y1.getId(), values(z.getId())),
+                        project(y2.getId(), values(z.getId()))));
+    }
+
+    private static void assertMatchesStructure(PlanNode actual, PlanNode expected)
+    {
+        assertEquals(actual.getClass(), expected.getClass());
+        assertEquals(actual.getId(), expected.getId());
+        assertEquals(actual.getSources().size(), expected.getSources().size());
+
+        for (int i = 0; i < actual.getSources().size(); i++) {
+            assertMatchesStructure(actual.getSources().get(i), expected.getSources().get(i));
+        }
+    }
+
+    private int getChildGroup(Memo memo, int group)
+    {
+        PlanNode node = memo.getNode(group);
+        GroupReference child = (GroupReference) node.getSources().get(0);
+
+        return child.getGroupId();
+    }
+
+    private JoinNode join(PlanNodeId id, PlanNode left, PlanNode right)
+    {
+        return new JoinNode(id, JoinNode.Type.INNER, left, right, ImmutableList.of(), Optional.empty(), Optional.empty(), Optional.empty());
+    }
+
+    private JoinNode join(PlanNode left, PlanNode right)
+    {
+        return join(idAllocator.getNextId(), left, right);
+    }
+
+    private ProjectNode project(PlanNodeId id, PlanNode source)
+    {
+        return new ProjectNode(id, source, Assignments.of());
+    }
+
+    private ProjectNode project(PlanNode source)
+    {
+        return project(idAllocator.getNextId(), source);
+    }
+
+    private FilterNode filter(PlanNodeId id, PlanNode source)
+    {
+        return new FilterNode(id, source, BooleanLiteral.TRUE_LITERAL);
+    }
+
+    private FilterNode filter(PlanNode source)
+    {
+        return filter(idAllocator.getNextId(), source);
+    }
+
+    private ValuesNode values(PlanNodeId id)
+    {
+        return new ValuesNode(id, ImmutableList.of(), ImmutableList.of());
+    }
+
+    private ValuesNode values()
```

Comment:
It's not a general purpose dsl. It's just for convenience to avoid sprinkling a bunch of `new ProjectNode(...)`, etc. in every test in this class. Also, I don't really care about it being filters or projections or values, since the purpose is just to test whether the Memo is breaking up and reassembly the plans correctly. If the plan node hierarchy weren't so rigid, I would've created specialized node types. Hmm.. now that I think about it, I may be able to do just that anyway, since the code doesn't use visitors.